### PR TITLE
Shortcut for Renaming object in the Objects List Panel

### PR DESCRIPTION
**Pull request by danired**
The shortcut default key is F2, it can be configured in the Shorcuts tab (Preferences dialog).

#### Major changes:

I moved this state from `ObjectsList/index.js` to `SceneEditor/index.js`.

```typescript
const [
  renamedObjectWithContext,
  setRenamedObjectWithContext,
] = React.useState<?ObjectWithContext>(null);
```

The reason is for using the state `renamedObjectWithContext` in both _ObjectsList_ and _ToolbarCommands_.

GDevelop - Rename screenshot

### CODE REVIEW

File: newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js
Comment (AlexandreSi) [inline newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js:235]: @D8H need your approval on this!

File: newIDE/app/src/KeyboardShortcuts/DefaultShortcuts.js
Comment (AlexandreSi) [inline newIDE/app/src/KeyboardShortcuts/DefaultShortcuts.js:42]: 💜

File: newIDE/app/src/ObjectsList/index.js
Comment (AlexandreSi) [inline newIDE/app/src/ObjectsList/index.js:234]: This line might need a comment to say that adding an object is always (at the moment) going to the scene (layout), and not to the project (as a global object) so the context is not global.

File: newIDE/app/src/ObjectsList/index.js
Comment (AlexandreSi) [reply newIDE/app/src/ObjectsList/index.js:128]: It would remove the awkward thing of sending the objectWithContext and its name coming from the same object

File: newIDE/app/src/SceneEditor/index.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/index.js:159]: I don't think you need this state variable. It's a props of `ObjectsList` that you set with a method of `SceneEditor`

File: newIDE/app/src/SceneEditor/Toolbar.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/Toolbar.js:45]: The toolbar only needs to know if an object can be renamed right now, so I think it only needs 2 props: `canRenameObject` and `onRenameObject` (this one is equivalent to your `startRenamingObject` but with the convention we try to respect in the project)

File: newIDE/app/src/SceneEditor/ToolbarCommands.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/ToolbarCommands.js:20]: Reusing the naming of the `Toolbar` props, it should be `canRenameObject` and `onRenameObject`.
Moreover, it should not contain a notion of scene here as you should be able to rename both a scene object and a global object

**Review by AlexandreSi — CHANGES_REQUESTED** [review]
Thank you for this PR, it looks like many people are waiting for this!
It still needs changes to be merged but you're on a good path!

**Comment by danired** [comment]
Hi @AlexandreSi thanks for your feedback!

I will look at it today or tomorrow.

File: newIDE/app/src/SceneEditor/index.js
Comment (danired) [reply newIDE/app/src/SceneEditor/index.js:159]: Hi @AlexandreSi ,

Sorry, I think I need it because I have this method used in ToolbarCommand:

```typescript
_startRenamingObject = () => {
  this._onEditRenameObject(this.state.selectedObjectWithContext);
};
```

`_onEditRenameObject` sets this state:

```typescript
this.setState(
  {
	renamedObjectWithContext: objectWithContext
  }
);
```

Setting the state `renamedObjectWithContext` activates the input in the Objects List so the user can write the new name.

If we leave the state only in the ObjectList, I can't see how I can change in from the SceneEditor/index.js. That was on of my big problems.

If you have any clue about this tell me and I will check it.

File: newIDE/app/src/ObjectsList/index.js
Comment (danired) [reply newIDE/app/src/ObjectsList/index.js]: Hi. I changed this. I added your text as a comment because you explained it good

File: newIDE/app/src/ObjectsList/index.js
Comment (danired) [reply newIDE/app/src/ObjectsList/index.js]: Done, now we only hace 1 parameter `ObjectWithContext`

File: newIDE/app/src/SceneEditor/Toolbar.js
Comment (danired) [reply newIDE/app/src/SceneEditor/Toolbar.js]: Done!

File: newIDE/app/src/SceneEditor/ToolbarCommands.js
Comment (danired) [reply newIDE/app/src/SceneEditor/ToolbarCommands.js]: Done!

**Comment by danired** [comment]
Hi @AlexandreSi , I finished the changes.

Only we wait for @D8H review and my doubts regarding the state properties

**Comment by danired** [comment]
I forgot to say that perhaps there are methods and states previous to my changes that can be refactored. But I tried to made the changes so I modified a low number of files.

But if you think I could refactor them tell me and I will check it.

File: newIDE/app/src/ObjectsList/index.js
Comment (AlexandreSi) [inline newIDE/app/src/ObjectsList/index.js:234]: Could you put this comment on multiple lines so that it's easy to read?
The number of characters before the formatter automatically wraps the line is 80.
You can use something similar (no need to be perfectly accurate)

File: newIDE/app/src/ObjectsList/index.js
Comment (AlexandreSi) [inline newIDE/app/src/ObjectsList/index.js:235]: I think the usual way to mention the name of the parameter is to use 
```suggestion
          onObjectSelected(/* objectWithContext */{ object, global: false });
```

File: newIDE/app/src/ObjectsList/index.js
Comment (AlexandreSi) [inline newIDE/app/src/ObjectsList/index.js:200]: Here, you replaced the state hooks with 2 props. So the component does not really know how stable those 2 props are.
To make sure `ObjectsList` always uses props that are up to date in its hooks, you must make sure that they are included in their dependencies.
In particular, the methods `editName`, `rename` and `renderObjectMenuTemplate` miss `setRenamedObjectWithContext` in their dependencies.
Note: Here, your code works well because the props do not change at each interaction within ObjectsList, but it might be the case in the future, in which case you would have seen errors.

File: newIDE/app/src/ObjectsList/index.js
Comment (danired) [reply newIDE/app/src/ObjectsList/index.js]: Hi,
Done and pushed.

File: newIDE/app/src/SceneEditor/index.js
Comment (danired) [reply newIDE/app/src/SceneEditor/index.js]: Good one there. Done and pushed.

File: newIDE/app/src/ObjectsList/index.js
Comment (danired) [reply newIDE/app/src/ObjectsList/index.js:200]: `renderObjectMenuTemplate` I think doesn't use `setRenamedObjectWithContext`, but instead uses `preferences.values.userShortcutMap,`

image

Do you prefer to add both dependencies in `renderObjectMenuTemplate` or just one of them?

Comment (danired) [reply newIDE/app/src/ObjectsList/index.js:200]: And just a doubt: My Visual Studio always shows lot of errors regarding classes not found, like these:

image

Any clue on how to make these errors dissapear?

**Comment by AlexandreSi** [comment]
Thanks for the updates!
I checked out your branch to test it, and it works well!
Just a few notes on the usability, I don't think it's due to your changes, but I think it might be cool to have those:
- I expected to cancel my rename when hitting the `Esc` key. It can be a complicated change since it would impact every `SortableVisualizedItemList`. I'll maybe try to change that later
- When hitting `Enter` to validate my rename, if I hit back `F2`, nothing happens because the item is unselected. I think it's due to `_onEditRenameObject` setting `selectedObjectWithContext` to `null`. It might be interesting to not set it to null in that case. What do you think?

Thanks again for the work!

**Comment by danired** [comment]
> * When hitting `Enter` to validate my rename, if I hit back `F2`, nothing happens because the item is unselected. I think it's due to `_onEditRenameObject` setting `selectedObjectWithContext` to `null`. It might be interesting to not set it to null in that case. What do you think?

Yes, although I would like that too, I kept the same functionality as its when choosing "Rename" option in the context menu.

I think both (menu option and shortcut) should do the same.

And about the `Esc` key, I like that too, and I thinked about it, but once again prevailed the current functionality in `master`.

Perhaps we could share work and you can check the `Esc` and I could see for the object remain selected.

**Comment by danired** [comment]
> * When hitting `Enter` to validate my rename, if I hit back `F2`, nothing happens because the item is unselected. I think it's due to `_onEditRenameObject` setting `selectedObjectWithContext` to `null`. It might be interesting to not set it to null in that case. What do you think?

Yes, although I would like that too, I prefered to keep the same functionality as its when choosing "_Rename_" option in the context menu. That is because I think both (menu option and shortcut) should do the same.

And about the `Esc` key, I like that too, and I thinked about it, but once again prevailed the current functionality in `master`.

Perhaps we could share work and you can check the `Esc` and I could see for the object remain selected.

File: newIDE/app/src/ObjectsList/index.js
Comment (danired) [reply newIDE/app/src/ObjectsList/index.js:200]: In the meanwhile I pushed the fix with the new props in the hooks

Comment (danired) [reply newIDE/app/src/ObjectsList/index.js:200]: > And just a doubt: My Visual Studio always shows lot of errors regarding classes not found, like these:
> 
> image
> 
> Any clue on how to make these errors dissapear?

I checked the readme and installed "Flow language support", now is ok!

Comment (AlexandreSi) [reply newIDE/app/src/ObjectsList/index.js:200]: Great you saw that the Flow plugin was needed! You're one step closer to an ideal development environment for GDevelop

**Comment by AlexandreSi** [comment]
I couldn't push the following changes to your repo. Could you use those changes and we should be good to go!

image

The first is a linting issue (the react build would have failed, it is shown in the react dev terminal normally).
The second is to keep the object selected after the rename to be able to hit F2 once again.
We'll see for the cancel action later

**Comment by danired** [comment]
I invited you to the repos, can you upload the changes now?

Or if you prefer I can add them myself.

File: newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js
Comment (AlexandreSi) [reply newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js:234]: From what I see, this is ok at the moment. It will be changed when the feature to rename children in EBOs is implemented.

**Review by AlexandreSi — APPROVED** [review]

File: newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js
Comment (danired) [reply newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js:234]: Good, thanks!

Sorry, what is "EBOs" ? 😅

File: newIDE/app/src/ObjectsList/index.js
Comment (ClementPasteau) [inline newIDE/app/src/ObjectsList/index.js:151]: I'm a bit confused by passing down props that are name `setXXX` because you don't really know when/why you should use them and what it does.
Should this rather be called something more specific?
Like `onRenameStart()` or `onRenameFinish()`.
It makes more sense from the component's point of view what it should do with it.

File: newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js
Comment (ClementPasteau) [inline newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js:168]: is that a good thing to keep the previous selection when selecting 1 element?
If that's on purpose, should the function be renamed to `_onAdditionalObjectSelected` ?

File: newIDE/app/src/SceneEditor/index.js
Comment (ClementPasteau) [inline newIDE/app/src/SceneEditor/index.js:468]: nit: we update "the" toolbar

File: newIDE/app/src/SceneEditor/index.js
Comment (ClementPasteau) [inline newIDE/app/src/SceneEditor/index.js:459]: same here, the logic changes to adding an additional object to the selection.
Is that on purpose?

**Review by ClementPasteau — COMMENTED** [review]
I've added a few comments / questions to the PR :)

File: newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js
Comment (ClementPasteau) [reply newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js:168]: EDIT: I've re-read the code and now understands this is the same :)

File: newIDE/app/src/SceneEditor/index.js
Comment (ClementPasteau) [reply newIDE/app/src/SceneEditor/index.js:459]: edit: same here, sorry :)

File: newIDE/app/src/ObjectsList/index.js
Comment (4ian) [reply newIDE/app/src/ObjectsList/index.js:151]: Agree that it should be called onXxx. or onSetXxx.
THis is because the children is not aware of what it does exactly, but the name gives an indication about "when" it must be called ("on" something happening...).
The parent is the one responsible for doing something.

File: newIDE/app/src/ObjectsList/index.js
Comment (4ian) [inline newIDE/app/src/ObjectsList/index.js:236]: Let's remove the /* objectWithContext */ or write `/* objectWithContext= */`

File: newIDE/app/src/ObjectsList/index.js
Comment (4ian) [inline newIDE/app/src/ObjectsList/index.js:234]:
```suggestion
          // For now, a new object is always add to the scene (layout)
```

File: newIDE/app/src/ObjectsList/index.js
Comment (4ian) [inline newIDE/app/src/ObjectsList/index.js:687]: We should probably NOT have a default value. The shortcut map should be the source of truth and not accelerator, apart from the non customisable one, should live in the code. Otherwise, what happen if the shortcut is not set? F2 would be displayed and would not work. That would be a bug.

File: newIDE/app/src/KeyboardShortcuts/DefaultShortcuts.js
Comment (4ian) [inline newIDE/app/src/KeyboardShortcuts/DefaultShortcuts.js:42]: Follow the convention of the other shortcuts by using a verb first:
```suggestion
  RENAME_SCENE_OBJECT: 'F2',
```

File: newIDE/app/src/SceneEditor/index.js
Comment (4ian) [inline newIDE/app/src/SceneEditor/index.js:691]:
```suggestion
  _startRenamingObject = (objectWithContext: ?ObjectWithContext) => {
```

File: newIDE/app/src/SceneEditor/index.js
Comment (4ian) [inline newIDE/app/src/SceneEditor/index.js:704]: Follow a more consistent naming for your functions. The name explains what the function does. If two functions do similar things, their names will follow the same pattern:

```suggestion
  _startRenamingSelectedObject = () => {
    this._startRenamingObject(this.state.selectedObjectWithContext);
```

File: newIDE/app/src/SceneEditor/index.js
Comment (4ian) [inline newIDE/app/src/SceneEditor/index.js:220]: We now unfortunately have two things:
`selectedObjectNames` and `selectedObjectWithContext`.

Which one is the source of truth? When do one is used and not the other? This is not clear because the good answer is: "this should be a single variable and not two".

This is not an easy task, but we need to replace one with the other. It's important because otherwise we start introducing two things doing the same thing, which means bugs and a hard to understand code for future developers.
I *think* we should replace `selectedObjectNames` by `selectedObjectsWithContext` (which would be an array of objects with contexts). 

Note that I'm maybe mistaken, but if that's the case, the names of the variables should be changed to make clear what is used for what :)

**Review by 4ian — COMMENTED** [review]
There is two state variables that express what object(s) is/are selected. This seems something that is very important to fix and make clean.

The rest is mostly comments on nitpickings/namings :) Great job for what you did already!

File: newIDE/app/src/ObjectsList/index.js
Comment (danired) [reply newIDE/app/src/ObjectsList/index.js:687]: I agree. FYI I only did that because I found the same thing in `EventSheet/index.js`.

I will remove it.

File: newIDE/app/src/SceneEditor/index.js
Comment (danired) [reply newIDE/app/src/SceneEditor/index.js:220]: When I began the task I thought the same but then I saw that I had to change too many lines and perhaps I was not allowed to do so many changes.

If you agree, I could remove the `selectedObjectNames` variable, that is only a array of strings, and use only `selectedObjectsWithContext` (or `selectedObjects`), that has more information.

**Comment by danired** [comment]
I made these changes and pushed. Basically it is all renaming:

- Rename `SCENE_OBJECT_RENAME` -> `RENAME_SCENE_OBJECT`
- ObjectsList -> `setRenamedObjectWithContext` -> `onRenameStart`
- ObjectsList -> `onRenameObject` -> `onRenameFinish`
- SceneEditor -> `_startRenamingObject` -> `_startRenamingSelectedObject`

I think it's pending only:

- What to do with `selectedObjectNames`

**Comment by AlexandreSi** [comment]
Sorry I did not spot that duplicated source of truth!
I think you can go with a `selectedObjectsWithContext` array.
It could be many lines indeed to change but I hope they are straightforward. I can help if needed.
I would just be careful when setting an object as global object (change of context) to make sure we don't keep outdated data.

**Comment by danired** [comment]
Done! It was easier than I thought...
In SceneEditor, I removed the array `selectedObjectNames` and only it uses `selectedObjects`.

For easy review I only changed `SceneEditor/index.js`. I saw that in the below files it's used a prop `selectedObjectNames` (_strings array_), but I don't know if I can change them too or could be too much.

- FullSizeInstancesEditorWithScrollbars
- ObjectsList

image

File: newIDE/app/src/SceneEditor/index.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/index.js:489]: Prefer using a consistent name. This should be `selectedObject`.

File: newIDE/app/src/SceneEditor/index.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/index.js:491]: It feels a bit weird to read this: `objectSelected.object`. I think you should go with the slightly heavier but clearer naming `selectedObjectsWithContext`

File: newIDE/app/src/SceneEditor/index.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/index.js:1039]: typo ~`ignoreselectedObjectsForContextMenu`~ `ignoreSelectedObjectsForContextMenu`

File: newIDE/app/src/SceneEditor/index.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/index.js:1119]: I think you can compute the object name at the start of the method and then use it in the items that need it

File: newIDE/app/src/SceneEditor/index.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/index.js:1440]: Same here, you can compute the array of object names at the start of the `render` method

File: newIDE/app/src/SceneEditor/index.js
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/index.js:546]: I'm a bit afraid of the efficiency of this computation:
- First you enumerate both the project and the layout objects
- Then, for each object, you search its presence in the instancesNames array
I think you can at least keep the `uniq` method to reduce the cost of the second step.
Also, it might be worth adding the possibility to filter objects by name in the `enumeratObjects` method.
Also, can you give a human readable name to the variable `e`? I thought it was supposed to represent an event at first.

**Comment by danired** [comment]
Hi @AlexandreSi I think it's all done. Could you check it please?

Thanks for your feedback!

**Comment by AlexandreSi** [comment]
I'll have a look on monday, thanks for the update!

**Review by ClementPasteau — APPROVED** [review]
lgtm 👍

File: newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js
Comment (4ian) [inline newIDE/app/src/EventsBasedObjectEditor/EventBasedObjectChildrenEditor.js:234]: This might be breaking the renaming of child objects in custom objects. Can you double check @ClementPasteau?
I think we need to add a state here.

File: newIDE/app/src/SceneEditor/index.js
Comment (4ian) [inline newIDE/app/src/SceneEditor/index.js:1493]: Nit, but important to be extra clear and consistent with `onObjectCreated`, `onEditrObject`...:
```suggestion
                onRenameObjectStart={this._onRenameObjectStart}
                onRenameObjectFinish={this._onRenameObjectFinish}
```

**Review by 4ian — CHANGES_REQUESTED** [review]

**Comment by danired** [comment]
Thank you all for your effort and review! I will test it tonight and check for improvements

**Pull request merged**
